mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala

```diff
@@ -341,11 +341,11 @@ class LogisticRegression @Since("1.2.0") (
                 regParamL1
               } else {
                 // If `standardization` is false, we still standardize the data
-                // to improve the rate of convergence; as a result, we have to
-                // perform this reverse standardization by penalizing each component
-                // differently to get effectively the same objective function when
+                // to improve the rate of convergence unless the standard deviation is zero;
+                // as a result, we have to perform this reverse standardization by penalizing
+                // each component differently to get effectively the same objective function when
                 // the training dataset is not standardized.
-                if (featuresStd(index) != 0.0) regParamL1 / featuresStd(index) else 0.0
+                if (featuresStd(index) != 0.0) regParamL1 / featuresStd(index) else regParamL1
```

Comment:
@coderxiang You may try it out. When `value` is very small like `1e-6` as a constant comparing the rest of the features, the corresponding coefficient will be very large. In this case, the optimization on coefficients will be on different scales, and this often causes some convergence issue in line search. 

Similar argument can be made when the constant `value` is very large comparing the rest of the features. 

That's why we still standardize the features even users ask `standardization == false`.